File: sdks/python/src/autonoma/handler.py
"""Request routing for discover/up/down protocol actions (Scenario v2).

``discover`` lists the registered scenarios; ``up`` looks a scenario up by
name, runs its free-form ``up``, signs a teardown token carrying the scenario
name, and responds; ``down`` recovers the scenario name from the verified
token and routes to that scenario's ``down``. There is no create-graph
interpreter and no factory-derived discover schema.
"""

from __future__ import annotations

import inspect
import json
import uuid
from typing import Any

from autonoma.errors import (
    AutonomaError,
    invalid_body,
    invalid_teardown_token,
    invalid_signature,
    same_secrets,
    unknown_action,
    unknown_environment,
)
from autonoma.hmac_util import verify_signature
from autonoma.refs import sign_refs, verify_refs
from autonoma.types import (
    HandlerConfig,
    HandlerRequest,
    HandlerResponse,
    ScenarioDefinition,
    ScenarioDownContext,
    ScenarioUpContext,
    ScenarioUpResult,
)

_DEFAULT_EXPIRES_IN_SECONDS = 3600


def _load_protocol_version() -> str:
    try:
        from pathlib import Path

        return (
            (Path(__file__).resolve().parents[4] / "protocol" / "version.txt")
            .read_text()
            .strip()
        )
    except (OSError, IndexError):
        return "2.0"


PROTOCOL_VERSION = _load_protocol_version()


def _build_sdk_meta(config: HandlerConfig) -> dict[str, Any]:
    sdk = config.sdk or {}
    return {
        "version": PROTOCOL_VERSION,
        "sdk": {
            "language": "python",
            "orm": sdk.get("orm", "unknown"),
            "server": sdk.get("server", "unknown"),
        },
    }


async def handle_request(config: HandlerConfig, req: HandlerRequest) -> HandlerResponse:
    """Handle an incoming Autonoma protocol request."""
    try:
        if config.shared_secret == config.signing_secret:
            raise same_secrets()

        signature: str = (
            req.headers.get("x-signature") or req.headers.get("X-Signature") or ""
        )
        if not verify_signature(req.body, signature, config.shared_secret):
            raise invalid_signature()

        try:
            body: dict[str, Any] = json.loads(req.body)
        except (json.JSONDecodeError, ValueError):
            raise invalid_body("invalid JSON")

        action = body.get("action")
        if not action:
            raise invalid_body(
                "missing action. expected one of 'discover', 'up' or 'down'"
            )

        if action == "discover":
            return await _handle_discover(config)
        if action == "up":
            return await _handle_up(config, body)
        if action == "down":
            return await _handle_down(config, body)
        raise unknown_action(action)

    except AutonomaError as e:
        return HandlerResponse(
            status=e.status, body={"error": e.message, "code": e.code}
        )
    except Exception as e:
        return HandlerResponse(
            status=500, body={"error": str(e), "code": "INTERNAL_ERROR"}
        )


# ---------------------------------------------------------------------------
# discover
# ---------------------------------------------------------------------------


async def _handle_discover(config: HandlerConfig) -> HandlerResponse:
    scenarios = [
        {"name": s.name, "description": s.description} for s in config.scenarios
    ]
    return HandlerResponse(
        status=200,
        body={**_build_sdk_meta(config), "scenarios": scenarios},
    )


# ---------------------------------------------------------------------------
# up
# ---------------------------------------------------------------------------


async def _handle_up(config: HandlerConfig, body: dict[str, Any]) -> HandlerResponse:
    name = _read_scenario_name(body)
    if name is None:
        raise invalid_body('missing "scenario.name" in request body')

    scenario = _find_scenario(config, name)
    if scenario is None:
        raise unknown_environment(name)

    test_run_id: str = body.get("testRunId") or str(uuid.uuid4())

    result = scenario.up(ScenarioUpContext(test_run_id=test_run_id))
    if inspect.isawaitable(result):
        result = await result

    auth, teardown = _unpack_up_result(result)

    teardown_token = sign_refs(
        {"refs": teardown or {}, "testRunId": test_run_id, "environment": name},
        config.signing_secret,
    )

    expires_in_seconds = (
        config.expires_in_seconds
        if config.expires_in_seconds is not None
        else _DEFAULT_EXPIRES_IN_SECONDS
    )

    response_body: dict[str, Any] = {**_build_sdk_meta(config)}
    if auth is not None:
        response_body["auth"] = auth
    response_body["teardownToken"] = teardown_token
    response_body["expiresInSeconds"] = expires_in_seconds

    return HandlerResponse(status=200, body=response_body)


def _unpack_up_result(
    result: Any,
) -> tuple[dict[str, Any] | None, dict[str, Any] | None]:
    """Accept either a ``ScenarioUpResult`` or a plain dict from a scenario's ``up``."""
    if result is None:
        return None, None
    if isinstance(result, ScenarioUpResult):
        return result.auth, result.teardown
    if isinstance(result, dict):
        return result.get("auth"), result.get("teardown")
    raise AutonomaError(
        "Scenario up() must return a dict or ScenarioUpResult with optional "
        "auth/teardown keys",
        "INTERNAL_ERROR",
        500,
    )


# ---------------------------------------------------------------------------
# down
# ---------------------------------------------------------------------------


async def _handle_down(config: HandlerConfig, body: dict[str, Any]) -> HandlerResponse:
    teardown_token = body.get("teardownToken")
    if not teardown_token:
        raise invalid_body("missing teardownToken")

    try:
        payload = verify_refs(teardown_token, config.signing_secret)
    except Exception as e:
        raise invalid_teardown_token(str(e))

    teardown: dict[str, Any] = payload.get("refs") or {}
    test_run_id: str = payload.get("testRunId", "")
    # The verified token is authoritative for routing; any scenario name on
    # the request body is ignored.
    name = payload.get("environment") or ""

    scenario = _find_scenario(config, name) if name else None
    if scenario is not None and scenario.down is not None:
        result = scenario.down(
            ScenarioDownContext(name=name, teardown=teardown, test_run_id=test_run_id)
        )
        if inspect.isawaitable(result):
            await result

    return HandlerResponse(status=200, body={**_build_sdk_meta(config), "ok": True})


# ---------------------------------------------------------------------------
# helpers
# ---------------------------------------------------------------------------


def _find_scenario(config: HandlerConfig, name: str) -> ScenarioDefinition | None:
    for scenario in config.scenarios:
        if scenario.name == name:
            return scenario
    return None


def _read_scenario_name(body: dict[str, Any]) -> str | None:
    scenario = body.get("scenario")
    if isinstance(scenario, dict):
        name = scenario.get("name")
        if isinstance(name, str):
            return name
    return None

File: sdks/ruby/lib/autonoma/handler.rb
# frozen_string_literal: true

require "json"
require "securerandom"

require_relative "hmac"
require_relative "refs"
require_relative "errors"
require_relative "types"

module Autonoma
  # Request routing for discover / up / down protocol actions (Scenario v2).
  #
  # discover lists the registered scenarios; up looks a scenario up by name,
  # runs its free-form up, signs a teardown token carrying the scenario name,
  # and responds; down recovers the scenario name from the verified token and
  # routes to that scenario's down. There is no create-graph interpreter and no
  # factory-derived discover schema.
  module Handler
    PROTOCOL_VERSION = begin
      File.read(File.expand_path("../../../../protocol/version.txt", __dir__)).strip
    rescue Errno::ENOENT, Errno::EACCES
      "2.0"
    end

    DEFAULT_EXPIRES_IN_SECONDS = 3600

    def self.handle_request(config, req)
      if config.shared_secret == config.signing_secret
        raise Errors.same_secrets
      end

      signature = req.headers["x-signature"] || req.headers["X-Signature"] || ""
      unless Hmac.verify_signature(req.body, signature, config.shared_secret)
        raise Errors.invalid_signature
      end

      begin
        body = JSON.parse(req.body)
      rescue JSON::ParserError
        raise Errors.invalid_body("invalid JSON")
      end

      action = body["action"]
      if action.nil? || action.to_s.empty?
        raise Errors.invalid_body('missing action. expected one of "discover", "up" or "down"')
      end

      case action
      when "discover"
        handle_discover(config)
      when "up"
        handle_up(config, body)
      when "down"
        handle_down(config, body)
      else
        raise Errors.unknown_action(action)
      end
    rescue AutonomaError => e
      HandlerResponse.new(status: e.status, body: { "error" => e.message, "code" => e.code })
    rescue StandardError => e
      HandlerResponse.new(status: 500, body: { "error" => e.message, "code" => "INTERNAL_ERROR" })
    end

    # -----------------------------------------------------------------------
    # discover
    # -----------------------------------------------------------------------

    def self.handle_discover(config)
      scenarios = (config.scenarios || []).map do |s|
        { "name" => s.name, "description" => s.description }
      end
      HandlerResponse.new(
        status: 200,
        body: build_sdk_meta(config).merge("scenarios" => scenarios)
      )
    end

    # -----------------------------------------------------------------------
    # up
    # -----------------------------------------------------------------------

    def self.handle_up(config, body)
      name = read_scenario_name(body)
      raise Errors.invalid_body('missing "scenario.name" in request body') if name.nil? || name.empty?

      scenario = find_scenario(config, name)
      raise Errors.unknown_environment(name) if scenario.nil?

      test_run_id = string_or_nil(body["testRunId"]) || SecureRandom.uuid

      result = scenario.up.call(ScenarioUpContext.new(test_run_id: test_run_id))
      result = {} if result.nil?

      auth = result_field(result, :auth)
      teardown = result_field(result, :teardown)

      teardown_token = Refs.sign_refs(
        { "refs" => teardown || {}, "testRunId" => test_run_id, "environment" => name },
        config.signing_secret
      )

      expires = config.expires_in_seconds || DEFAULT_EXPIRES_IN_SECONDS

      resp = build_sdk_meta(config)
      resp["auth"] = auth unless auth.nil?
      resp["teardownToken"] = teardown_token
      resp["expiresInSeconds"] = expires

      HandlerResponse.new(status: 200, body: resp)
    end

    # -----------------------------------------------------------------------
    # down
    # -----------------------------------------------------------------------

    def self.handle_down(config, body)
      teardown_token = string_or_nil(body["teardownToken"])
      raise Errors.invalid_body("missing teardownToken") if teardown_token.nil? || teardown_token.empty?

      begin
        payload = Refs.verify_refs(teardown_token, config.signing_secret)
      rescue StandardError => e
        raise Errors.invalid_teardown_token(e.message)
      end

      teardown = payload["refs"] || {}
      test_run_id = payload["testRunId"] || ""
      # The verified token is authoritative for routing; any scenario name on
      # the request body is ignored.
      name = payload["environment"] || ""

      unless name.empty?
        scenario = find_scenario(config, name)
        if scenario && scenario.down
          scenario.down.call(
            ScenarioDownContext.new(name: name, teardown: teardown, test_run_id: test_run_id)
          )
        end
      end

      HandlerResponse.new(status: 200, body: build_sdk_meta(config).merge("ok" => true))
    end

    # -----------------------------------------------------------------------
    # helpers
    # -----------------------------------------------------------------------

    def self.build_sdk_meta(config)
      sdk = config.sdk || {}
      {
        "version" => PROTOCOL_VERSION,
        "sdk" => {
          "language" => "ruby",
          "orm" => sdk["orm"] || sdk[:orm] || "unknown",
          "server" => sdk["server"] || sdk[:server] || "unknown"
        }
      }
    end

    def self.find_scenario(config, name)
      (config.scenarios || []).find { |s| s.name == name }
    end

    # Read body.scenario.name from an untrusted JSON body.
    def self.read_scenario_name(body)
      scenario = body["scenario"]
      return nil unless scenario.is_a?(Hash)

      name = scenario["name"]
      name.is_a?(String) ? name : nil
    end

    def self.string_or_nil(value)
      value.is_a?(String) ? value : nil
    end

    # Read a field from a scenario up result that may be a ScenarioUpResult
    # struct or a plain Hash keyed by symbols or strings.
    def self.result_field(result, key)
      if result.is_a?(Hash)
        result[key].nil? ? result[key.to_s] : result[key]
      elsif result.respond_to?(key)
        result.public_send(key)
      end
    end

    private_class_method :build_sdk_meta, :handle_discover, :handle_up,
                         :handle_down, :find_scenario, :read_scenario_name,
                         :string_or_nil, :result_field
  end
end
